include/aspect/particle/property/cpo.h

```diff
@@ -0,0 +1,361 @@
+                                       std::vector<std::vector<Tensor<2,3> > > &a_cosine_matrices_mineral,
+                                       std::vector<std::vector<double> > &volume_fractions_mineral_derivatives,
+                                       std::vector<std::vector<Tensor<2,3> > > &a_cosine_matrices_mineral_derivatives) const;
+
+          /**
+           * Updates the volume fractions and rotation matrices with a Forward Euler scheme.
+           */
+          double
+          advect_forward_euler(std::vector<double> &volume_fractions,
+                               std::vector<Tensor<2,3> > &a_cosine_matrices,
+                               const std::pair<std::vector<double>, std::vector<Tensor<2,3> > > &derivatives,
+                               const double strain_rate_second_invariant,
+                               const double dt) const;
+          /**
+           * Updates the volume fractions and rotation matrices with a Backward Euler scheme.
+           */
+          double
+          advect_backward_euler(std::vector<double> &volume_fractions,
+                                std::vector<Tensor<2,3> > &a_cosine_matrices,
+                                const std::pair<std::vector<double>, std::vector<Tensor<2,3> > > &derivatives,
+                                const double strain_rate_second_invariant,
+                                const double dt) const;
+
+          /**
+           * Updates the volume fractions and rotation matrices with a Crank-Nicolson scheme.
+           */
+          double
+          advect_Crank_Nicolson(std::vector<double> &volume_fractions,
+                                std::vector<Tensor<2,3> > &a_cosine_matrices,
+                                const std::pair<std::vector<double>, std::vector<Tensor<2,3> > > &derivatives,
+                                std::vector<double> &previous_volume_fraction_derivatives,
+                                std::vector<Tensor<2,3> > &previous_a_cosine_matrices_derivatives,
+                                const double strain_rate_second_invariant,
+                                const double dt) const;
+
+
+          /**
+           * A interface function which delegates the computation the derivatives
+           * to specialized functions.
+           */
+          std::pair<std::vector<double>, std::vector<Tensor<2,3> > >
+          compute_derivatives(const std::vector<double> &volume_fractions,
+                              const std::vector<Tensor<2,3> > &a_cosine_matrices,
+                              const SymmetricTensor<2,3> &strain_rate_nondimensional,
+                              const Tensor<2,3> &velocity_gradient_tensor_nondimensional,
+                              const double volume_fraction_mineral,
+                              const std::array<double,4> &ref_resolved_shear_stress) const;
+
+          /**
+           * Computes the update for the property rotating passively with the particle.
```

Comment:
Hmm, so it is the time derivative you are computing here? Please specify. And which property? According to the function name the `spin_tensor`? But which one is that, I do not think this was mentioned before. Also explain the parameters, at least the ones not explained in the general class documentation. And what does the function return (and why are those vectors)?